pkg/operator/csi/csidefaultstorageclasscontroller/csi_default_storageclass_controller.go

```diff
@@ -0,0 +1,135 @@
+)
+
+// This Controller deploys a StorageClass provided by CSI driver operator
+// and decides if this StorageClass should be applied as default - if requested.
+// If operator wants to request it's StorageClass to be created as default,
+// the asset file provided to this controller must have defaultScAnnotationKey set to "true".
+// Based on the current StorageClasses in the cluster the controller can decide not
+// to deploy given StorageClass as default if there is already an existing default.
+// Also, the controller will emit alerts if more than one StorageClass defined
+// as default exists - this is not a recommended setup.
+// It produces following Conditions:
+// DefaultStorageClassControllerDegraded - failed to apply StorageClass provided
+type CSIDefaultStorageClassController struct {
+	name               string
+	assetFunc          resourceapply.AssetFunc
+	file               string
+	kubeClient         kubernetes.Interface
+	storageClassLister v1.StorageClassLister
+	operatorClient     v1helpers.OperatorClient
+	eventRecorder      events.Recorder
+}
+
+func NewCSIDefaultStorageClassController(
+	name string,
+	assetFunc resourceapply.AssetFunc,
+	file string,
+	kubeClient kubernetes.Interface,
+	InformerFactory informers.SharedInformerFactory,
+	operatorClient v1helpers.OperatorClient,
+	eventRecorder events.Recorder) factory.Controller {
+	c := &CSIDefaultStorageClassController{
+		name:               name,
+		assetFunc:          assetFunc,
+		file:               file,
+		kubeClient:         kubeClient,
+		storageClassLister: InformerFactory.Storage().V1().StorageClasses().Lister(),
+		operatorClient:     operatorClient,
+		eventRecorder:      eventRecorder,
+	}
+
+	return factory.New().WithSync(
+		c.Sync,
+	).ResyncEvery(
+		time.Minute,
+	).WithSyncDegradedOnError(
+		operatorClient,
+	).WithInformers(
+		operatorClient.Informer(),
+		InformerFactory.Storage().V1().StorageClasses().Informer(),
+	).ToController(
+		"DefaultStorageClassController",
+		eventRecorder,
+	)
+}
+
+func (c *CSIDefaultStorageClassController) Sync(ctx context.Context, syncCtx factory.SyncContext) error {
+	klog.V(4).Infof("DefaultStorageClassController sync started")
+	defer klog.V(4).Infof("DefaultStorageClassController sync finished")
+
+	opSpec, _, _, err := c.operatorClient.GetOperatorState()
+	if err != nil {
+		return err
+	}
+	if opSpec.ManagementState != operatorapi.Managed {
+		return nil
+	}
+
+	syncErr := c.syncStorageClass(ctx)
+
+	return syncErr
+}
+
+func (c *CSIDefaultStorageClassController) syncStorageClass(ctx context.Context) error {
+	expectedScBytes, err := c.assetFunc(c.file)
+	if err != nil {
+		return err
+	}
+
+	expectedSC := resourceread.ReadStorageClassV1OrDie(expectedScBytes)
+
+	existingSCs, err := c.storageClassLister.List(labels.Everything())
+	if err != nil {
+		klog.V(2).Infof("could not list StorageClass objects")
+		return err
+	}
+
+	defaultScCount := 0
+	expectedScNonDefaultExists := false
```

Comment:
nit: `defaultSCCount`, `expectedSCNonDefaultExists` (SC is an abbreviation -> all uppercase, unless it's the first word in an identifier)

Reply:
done